lib/core/src/Cardano/Pool/Metrics.hs

```diff
@@ -52,8 +73,105 @@ import GHC.Generics
 import Numeric.Natural
     ( Natural )
 
+import qualified Data.List.NonEmpty as NE
 import qualified Data.Map.Strict as Map
 
+data Block = Block
+    { header :: BlockHeader
+    , producer :: Maybe PoolId
+    } deriving (Eq, Show, Generic)
+
+data ErrMetricsWorkerErr
+    = ErrMetricsWorkerErrTipTooVolatile
+    | ErrMetricsWorkerStakeDistributionUnavailible
+    | ErrMetricsWorkerDataInconsistencyBlockWithoutProducer
+    | ErrMetricsWorkerDataInconsistencyDuplicateBlock
+    | ErrMetricsWorkerTipIsUnreachable
+    deriving (Show, Eq)
+
+-- | @worker@ follows the chain and puts pool productions and stake
+-- distributions to a @DBLayer@, such that the data in the @DB@ always is
+-- consistent.
+--
+-- The pool productions and stake distrubtions in the DB can /never/ be from
+-- different forks.
+worker
+    :: NetworkLayer IO t Block
+    -> DBLayer IO
+    -> Trace IO Text
+    -> IO ()
+worker nl db tr = do
+    -- Read the k latest headers the DB knows about
+    let (_block0, bp) = staticBlockchainParameters nl
+    let k = fromIntegral . getQuantity . view #getEpochStability $ bp
+    knownHeaders <- readCursor db k
+
+    follow nl tr knownHeaders advance' rollback getHeader
+  where
+
+    advance'
+        :: NE.NonEmpty Block
+        -> BlockHeader
+        -> IO (FollowAction ErrMetricsWorkerErr)
+    advance' bs h = runExceptT (advance bs h) >>= return . \case
+        Left ErrMetricsWorkerErrTipTooVolatile
+            -> Retry
+        Left e
+            -> ExitWith e
+            -- TODO: Make all cases explicit here.
+        Right ()
+            -> Continue
+
+    advance
+        :: NE.NonEmpty Block
+        -> BlockHeader
+        -> ExceptT ErrMetricsWorkerErr IO ()
+    advance blocks nodeTip = do
+        -- Get a stake distribution and next batch of blocks such that we know
+        -- they are from the same chain.
+        --
+        -- Assumtion: We don't expect the node to switch from tip A to tip B,
+        -- and back to tip A.
+        --
+        -- If the tip is still the same after we have fetched the stake
+        -- distribution, we conclude that the stake-distrubtion is from the same
+        -- chain as the tip.
+        consistentDistr <- withinSameTip retryPolicy liftE getTip $ \nodeTip2 -> do
```

Comment:
what about using guards here?

```hs
\nodeTip2
    | nodeTip == nodeTip2 -> throwE ErrMetricsWorkeTipTooVolatile 
    | otherwise -> do
         ....
```

Reply:
Although, we probably want to retry here... Perhaps it's best to amend `withinSameTip` to that it takes an initial tip as argument!

Reply:
Good point… But actually… Maybe the whole retrying in `withinSameTip` is pointless.

We care about `stakeDistribution nl` being called while the the tip is still `nodeTip` (from the follow-callback). It should be enough to:
```haskell
res <- action
after <- getTip
guard (after == nodeTip)
``` 
i.e not fetching an extra tip before running the action. 

Furthermore: if the comparison fails, retrying won't help. (unless the node switches from A to B and back to A again, which we already assumed it cannot)

Reply:
8c04a4d

Reply:
But _maybe_ we should then also
- Remove the possibility of `Retry` in `FollowAction` (because we will never retry)
- Consider allowing callbacks of `[]` such that we can fetch stake distribution even if there are no new blocks. What if the node has stopped syncing with the network? We should still be able to fetch the stake-distribution.